### ISSUE TITLE
Let us write iOS-specific JS in TypeScript

### ISSUE THREAD

**Issue by lazerwalker** [flappy-royale/flappy-royale#51]
Currently, there's an `iOS/FlappyRoyale/JS` folder, and some corresponding hand-authored typings files in the main TS repo.

Since we'll have more iOS/JS integrations in the future (e.g. analytics, maybe ads), would be awesome if we could write that code in TS instead of JS.

I'm imagining the main `src/` folder has a folder of iOS-specific TS, and webpack is responsible for (a) dumping out type definition files and (b) putting the compiled JS in a place where Xcode can pick it up (either putting it directly in `iOS/FlappyRoyale/JS`, or we just update the folder ref in Xcode to be wherever)

**Comment by orta** [comment]
Interesting, I'd like to try keep feature parity with android too so maybe the can make that easier

**Comment by lazerwalker** [comment]
Smash cut to five years later, when we're stuck maintaining the most popular PhoneGap/Cordova replacement ;)

I haven't done any webview interop on Android, but looking at the docs it's poooosible we'll be able to get away without having to write any Android-specific JS. The kluge is necessary on iOS since WKWebView being out-of-process means we don't have access to JavaScriptCore-style JSContext interop.

**Comment by orta** [comment]
Great!

**Comment by lazerwalker** [comment]
Thinking about it, the easiest way to do this is have all of this code in the main app, but gated behind an e.g. `window.iOS` flag that gets set by Swift. 

This won’t work if we need the iOS-specific JS to template in Swift values, but our current solution also fails that. 

This adds sliiiightly more unused code to the app bundle for non-native-iOS clients, but also has the benefit that we can update that iOS code if need be without an App Store deploy.

**Comment by lazerwalker** [comment]
tl;dr: 

* The iOS app doesn't inject any JS other than a `window.isAppleApp` flag (presumably we can also add a corresponding `window.isAndroidApp` flag as well)
* The `haptics` object (in `haptics.ts`) has a implementation that uses the web Vibration API, but also overwrites itself with native functionality if `window.isAppleApp` is true
